### PR TITLE
Finished challenge

### PR DESCRIPTION
**Pull request by jbkimble**

### CODE REVIEW

**Comment by jcasimir** [comment]
Git is hard. 👍

File: well_formedness.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+require 'pry'
+
+class Validator
+  def validate(string)
+    result = get_relevant_chars(string)
+    return false if result.length.odd?
+    check_chars_match(result)
+
+
```
Comment (mikedao) [inline well_formedness.rb:9]: Watch your whitespace!  😸

File: well_formedness.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  def get_relevant_chars(string)
+    result = []
+    string.each_char do |char|
```
Comment (mikedao) [inline well_formedness.rb:14]: Which do you think is more efficient, using `each_char` with a collector local variable, or splitting the string and then mapping over it?

File: well_formedness.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+         char == ']'
+        result << char
+      end
+    end
+    return result
+  end
+
+  def check_chars_match(char_array)
+    until char_array.length == 0
+      if (char_array[0] == '(' && char_array[1] == ')') ||
```
Comment (mikedao) [inline well_formedness.rb:29]: I recommend pulling complex conditionals out into predicate methods.

File: well_formedness_spec.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    result = x.validate("()[")
+    assert_equal result, false
+  end
+
+  def test_it_fails_a_medium_example
+    x = Validator.new
+    result = x.validate("([)]")
+    assert_equal result, false
+  end
+
```
Comment (mikedao) [inline well_formedness_spec.rb:30]: Remember to throw in some sad path and edge tests too! Otherwise, well done!  👍

**Review by mikedao — APPROVED** [review]
I left some questions in here for you to think about.

**Pull request closed without merging**